internal/xfn/cached/cached_runner.go

```diff
@@ -0,0 +1,387 @@
+		log.Info("RunFunctionResponse cache miss", "reason", ReasonError, "err", err)
+		r.metrics.Miss(name, ReasonError)
+		r.metrics.Error(name)
+		return r.CacheFunction(ctx, name, req)
+	}
+
+	deadline := time.Unix(int64(binary.LittleEndian.Uint64(header)), 0) //nolint:gosec // False positive for G115.
+
+	// Cached response has expired.
+	if time.Now().After(deadline) {
+		log.Debug("RunFunctionResponse cache miss", "reason", ReasonDeadlineExpired, "deadline", deadline)
+		r.metrics.Miss(name, ReasonDeadlineExpired)
+		return r.CacheFunction(ctx, name, req)
+	}
+
+	// Everything after the first 8 bytes is our cached response.
+	msg, err := io.ReadAll(f)
+	if err != nil {
+		log.Info("RunFunctionResponse cache miss", "reason", ReasonError, "err", err)
+		r.metrics.Miss(name, ReasonError)
+		r.metrics.Error(name)
+		return r.CacheFunction(ctx, name, req)
+	}
+
+	// Our cached file is just 8 bytes of deadline, with no response. This
+	// is unlikely but if it did happen proto.Unmarshal would pass without
+	// error and we'd return an empty response.
+	if len(msg) == 0 {
+		log.Debug("RunFunctionResponse cache miss", "reason", ReasonEmptyResponse)
+		r.metrics.Miss(name, ReasonEmptyResponse)
+		return r.CacheFunction(ctx, name, req)
+	}
+
+	rsp := &fnv1.RunFunctionResponse{}
+	if err := proto.Unmarshal(msg, rsp); err != nil {
+		log.Info("RunFunctionResponse cache miss", "reason", ReasonError, "err", err)
+		r.metrics.Miss(name, ReasonError)
+		r.metrics.Error(name)
+		return r.CacheFunction(ctx, name, req)
+	}
+
+	log.Debug("RunFunctionResponse cache hit")
+	r.metrics.Hit(name)
+	r.metrics.ReadDuration(name, time.Since(start))
+	return rsp, nil
+}
+
+// CacheFunction runs a function and caches its response if the TTL is non-zero.
+func (r *FileBackedRunner) CacheFunction(ctx context.Context, name string, req *fnv1.RunFunctionRequest) (*fnv1.RunFunctionResponse, error) {
+	// If we don't have a cache key we can't cache the response. Just send
+	// it on. This should never happen.
+	if req.GetMeta().GetTag() == "" {
+		return r.wrapped.RunFunction(ctx, name, req)
+	}
+
+	key := filepath.Join(name, req.GetMeta().GetTag())
+	log := r.log.WithValues("name", name, "cache-key", key)
+
+	rsp, err := r.wrapped.RunFunction(ctx, name, req)
+	if err != nil {
+		return rsp, err
+	}
+
+	// Start timing the cache write _after_ we make the request.
+	start := time.Now()
+
+	ttl := rsp.GetMeta().GetTtl().AsDuration()
+	if ttl == 0 {
+		return rsp, nil
+	}
+
+	// Not all filesystems have btime, and Go doesn't expose a simple
+	// interface to get it. We just write our own at the beginning of the
+	// file. We use basic binary encoding (as opposed to gob or protobuf)
+	// because the fixed length makes it easy to slice the file into the
+	// timestamp header and the protobuf message body.
+	deadline := time.Now().Add(ttl)
+	header := make([]byte, 8)
+	binary.LittleEndian.PutUint64(header, uint64(time.Now().Add(ttl).Unix())) //nolint:gosec // False positive for G115.
```

Comment:
Shouldn't this use `deadline.Unix()` rather than re-calculating? There's a chance the written deadline wouldn't match what's logged below.

Reply:
Yes - not sure what happened here. I suspect I started refactoring `deadline` out into its own var but missed this.